### ISSUE TITLE
cholesky update/downdate?

### ISSUE THREAD

**Issue by ashimb9** [scipy/scipy#8188]
Is there a Scipy function to perform rank 1 cholesky updates and downdates (as in [MATLAB's cholupdate()](https://www.mathworks.com/help/matlab/ref/cholupdate.html))? I was unable to find anything in the docs and also here in the issues, so I would really appreciate any information. Thanks!

PS: If this has not been implemented here, I would have to implement it for my use-case and if contributions on such a functionality would be welcome, I would be happy to make a contribution here (although I would be limited to python/cython).

**Comment by ilayn** [comment]
[LAPACK did not include](http://icl.cs.utk.edu/lapack-forum/viewtopic.php?f=2&t=2646) the code that probably(!) matlab uses from LINPACK since nobody actually polished that code. There is Daniel Kressner's `QRUPDATE` package but I think the license is "free for noncommercial" and Octave is linking to that. 

https://sourceforge.net/projects/qrupdate/

I am not sure if this is actually Kressner's qrupdate. 

https://anchp.epfl.ch/QRupdate

There are other people who wrote similar functions but I couldn't find any benchmarks or stability analysis of these code. So I can't assess which way is to go. Hence it is indeed a bit cumbersome how to go about this. Maybe it is a better strategy to first consult this to [LAPACK people](https://github.com/Reference-LAPACK/lapack). If they mention that there is no interest for them to include then we can think of alternatives.

**Comment by josef-pkt** [comment]
Can this be computed from the qr_update?

Like qr_update and similar qr functions, I think including more linalg decomposition update functions would be useful to have in scipy.

**Comment by ilayn** [comment]
@ewmoore Care to chime in?

**Comment by xoviat** [comment]
xref Reference-LAPACK/lapack#209.

**Comment by ilayn** [comment]
@xoviat matlab uses lapack (linpack actually) already for that.

**Comment by xoviat** [comment]
Yes, it uses linpack but not lapack. So we should open an issue so that LAPACK will add the function, and then can backport it for older LAPACK APIs.

**Comment by ashimb9** [comment]
Thank you everyone for your very helpful comments (and please forgive my delayed response). For what it's worth, I completely agree that the ideal solution moving forward would be to have it integrated into LAPACK itself. Hopefully there will be positive news coming from that direction.

**Comment by ashimb9** [comment]
Hi everyone, so it has been almost 2 weeks since @xoviat posted the issue on the LAPACK discussion board but nobody over there has responded. I am not too familiar with the development process at LAPACK, so I am wondering if anyone knows if this is "normal"?

PS: My apologies for forgetting to note this earlier, but thank you @xoviat for taking the time to post on their issues board.

**Comment by josef-pkt** [comment]
some Lapack issues/question about adding QR or Cholesky rank-one update are 7 to 10 years old. 
I wouldn't plan on that they get magically added soon.

**Comment by ilayn** [comment]
Not only that but also even if it is added tomorrow to LAPACK, we won't be able to use it since we are barely trying to upgrade to LAPACK v3.4.1. So that would be at least quite a few years until we make it to v3.9.0. 

The reason that I've mentioned LAPACK however is because I wanted to get their opinion about the reference implementation so that we can mimick it. There are many versions lying around and I don't know which one is the reference implementation.

**Comment by xoviat** [comment]
The function won't be added to LAPACK unless someone submits a PR to them.
If no one takes the initiative, then it won't happen. Regardless, when is
the next SciPy release? We have plenty of time for this to be added.
Although I'm skeptical that it will make it in (even noting the length of
time) because no one with the capability to tackle this issue seems to have
much interest.

**Comment by xoviat** [comment]
@ilayn That's true, but it would be trivial to backport that particular
function (e.g. only include it if the symbol is undefined).

**Comment by ashimb9** [comment]
Of course. Although I was not expecting that it would be added to lapack anytime soon, my question was more in the vein of "are they not interested at all?". Having said that, I definitely was not aware that it would take "at least quite a few years" to get to the latest LAPACK version -- that is good to know.

**Comment by josef-pkt** [comment]
My impression as an outsider is that there is no reference implementation at LAPACK standards. At least for the QR update discussion I got the impression that there are unresolved numerical/technical details.
(In QR one problem is numerical stability in downdating when matrix is badly conditioned and what's the best way around it.)

**Comment by xoviat** [comment]
Let me be clear: if the function is added to LAPACK, then you will see it
in SciPy by the next feature release (1.10 as of now), assuming that it's
added with a reasonable lead-time to have it in SciPy. I can make sure of
that (assuming there are no objections, which I don't think there would
be). But if the function is not added to LAPACK, then you will probably not
see it added to SciPy because we need to be sure that it is numerically
stable as well.

**Comment by ilayn** [comment]
@ashimb9 LAPACK is like firmware, you change a single comma and people will start complaining about checksums not matching :) It's one of "the" most central libraries so people often tend to follow from a few versions behind. 

What I meant is if we know the reference algorithm, we can Cythonize it ourselves. But as I said, so far it is not clear what the update procedure is (at least to me). That's why I've pinged @ewmoore (let's try again one more time) because qr_update is authored by him in Cython and not in LAPACK routines.

**Comment by josef-pkt** [comment]
On a more optimistic note:
the previously linked http://icl.cs.utk.edu/lapack-forum/viewtopic.php?f=2&t=2646 
contains the code. dchdr1.txt  and dchur1.txt

AFAIK, the numerical standard in scipy is lower than in LAPACK. So, I don't think there should be a strong objection to having something like what's there, even if **eventually** it might be possible to replace it with a better LAPACK standard version.
That is, scipy could gain something soon that works for most cases even if it's not the best possible solution for ill conditioned cases.

"The downdating problem needs care." (quote from LAPACK issue) From what I read during the QR updating issue, QR updating is numerically stable, only the down dating might need a warning for numerical stability in ill-conditioned cases.

**Comment by ashimb9** [comment]
@josef-pkt Thanks for sharing your experience with the QR update discussion. It helps explain the apprehension on their side re the updates (or specifically the downdates). It would make sense for them hold back on implementation if the instability issues have not been appropriately resolved.

@xoviat I hear you. And thank you for your enthusiasm to bring this functionality to SciPy!

@ilayn Got it. And I agree, it does make a lot of sense to stick to the tried and tested versions, especially when a lot of critical functionality depends on it. Also, I hope my previous comment did not come off as a negative reaction -- FWIW that was certainly not the case.

**Comment by ilayn** [comment]
@ashimb9 No problem at all. Just tried to summarize the intertwined story. Took me a while to figure it out too.

**Comment by ewmoore** [comment]
Adding these routines is certainly in scope.  Like anything else, it requires someone to step up and write the code.  You can find plenty of papers that document algorithms.  I don't know which paper we should follow, an interested person needs to find these papers and review them, etc.  AFAIK this only become complex when downdating would cause the matrix to no longer be positive definite. 

We shouldn't wait for these to appear in lapack.  It would be nice certainly, but that might also mean waiting forever. One data point for not waiting would be that Julia appears to have merged rank-one updates / downdates in this ticket: Julialang/Julia#14243

**Comment by xoviat** [comment]
Note that I am NOT suggesting that we wait for these to appear in LAPACK. What I am suggesting is that IF someone wants to step up to write the code (I currently don't have the time), it would be more time-efficient for them to submit a PR to LAPACK and then simply copy the FORTRAN file into SciPy. Then, we would get LAPACK vendors to optimize the code for us.

**Comment by ilayn** [comment]
Thanks @ewmoore I'll have a look at them to see what I can do. In the meantime, I've found [DOI:10.1007/BF01933218](https://dx.doi.org/10.1007/BF01933218) as the starting point. Julia implementation looks a bit naive with manual Givens rotations. I'll try to make an initial PR soon.

**Comment by ashimb9** [comment]
@ilayn Also please feel free to let me know if I can be of any help at any point. And even though I am not familiar with Fortran, if there emerges anything python/cython related I can help with, I would be happy to do so.

**Comment by ilayn** [comment]
@ashimb9 Thanks. The paper I've linked to is from the second author of Alexander et al. There is also a comparison within the paper about that. That's why I've picked it up.

**Comment by ilayn** [comment]
OK here is a basic prototype with very little attention to optimization
```
from numpy import zeros_like, empty_like, sqrt, spacing
from scipy.linalg import LinAlgError


def algoB(R, z, lower=False):
    """
    Cholesky rank-1 downdate algorithm taken from [1]_ such that
    ``D'D = R'R - zz'``

    Parameters
    ----------
    R : (N, N) ndarray
        The 2D square array from which the triangular part will be used to
        read the Cholesky factor.The remaining parts are ignored.
    z : (N,) ndarray
        Downdate array
    lower: bool, optional
        Whether Cholesky factor is upper or lower triangular Cholesky
        factorization.  Default is upper-triangular.
    overwrite_R: bool, optional
        If set to True ....
    overwrite_z : bool, optional
        If set to True the entries of the input array will be modified during
        the computations instead of creating a new array.

    Returns
    -------
    D : (N, N) ndarray
        The resulting downdated Cholesky factor.

    References
    ----------
    .. [1] DOI:10.1007/BF01933218
    """
    # Just assuming arrayness, shape checks and overwrites etc. are handled.
    # Then... (with overwriting z)

    n = R.shape[0]
    eps = n * spacing(1.)  # For complex this needs modification
    alpha, beta = empty_like(z), empty_like(z)
    alpha[-1], beta[-1] = 1., 1.
    D = R.copy()

    for r in range(n):
        a = z[r] / R[r, r]
        alpha[r] = alpha[r-1] - a ** 2
        # Numerically zero or negative
        if alpha[r] < eps:
            # Made up err msg.
            raise LinAlgError('The Cholesky factor becomes nonpositive'
                              'with this downdate at the step {}'.format(r))
        beta[r] = sqrt(alpha[r])
        z[r+1:] -= a*R[r, r+1:]
        D[r, r:] *= beta[r] / beta[r-1]
        D[r, r+1:] -= a/(beta[r] * beta[r-1])*z[r+1:]

    return D
```
With this so far any array size worked (tested up to 5K) and only 4x slower than matlab.
```
n = 1000
A = np.random.rand(n, n)
A = A + A.T + n*np.eye(n)
R = sp.linalg.cholesky(A)
z = np.random.rand(n)
# Uncomment the following to make it occasionally trip up
#z *= 0.95 * np.sqrt(n)
zz = z.copy() # z is overwritten
D = algoB(R, zz)
np.testing.assert_allclose(D.T@ D - (R.T @ R - np.outer(z, z)),
                           np.zeros([n, n]), atol=n*np.spacing(100.))
```

The parts that need attention 
 - the walking backwards from bottom right corner up if `lower` is True.
 - Complex arrays and bookkeeping the datatypes to avoid accidental upcasting.
 - matlab seems to report the `r` value where it trips as an optional output. Should we? We try to avoid varying number of outputs.

@ashimb9 Would this be sufficient input for you stress test it and expand? If you need it, I can do the backwards walking too without too much hassle (famous last words).

**Comment by xoviat** [comment]
Hey, license compatible code here: https://github.com/mp4096/cholesky-update?files=1

**Comment by ilayn** [comment]
@xoviat Still using home-baked norms and givens rotations. We need to convert them into LAPACK calls. So again needs work. I don't mind but seems unnecessarily bloated.

**Comment by ashimb9** [comment]
@ilayn Thank you! I will give it a shot and get back to you. Also, since you mentioned comparison to Matlab performance, I am wondering if there are any unofficial thresholds on relative performance?

**Comment by ilayn** [comment]
@ashimb9 What kind of thresholds?

**Comment by ashimb9** [comment]
Threshold as in the max performance gap relative to Matlab. So you mentioned that it is currently 4x slower than Matlab right now, so I am wondering if "at the most" it can be, say, no more than 2X slower or something like that? Of course ideally we want it faster, but I am just trying to get a sense of the ceiling.

**Comment by ilayn** [comment]
I don't know any such thresholds but of course it should be reasonably fast. 4x is OKish for pure python if there is no other way. But matching matlab speed is a moving post with JIT and etc. 

It's a nice goal but the actual target is as fast as possible while keeping an eye on memory cost. Stability and maintainability are, in certain cases, more preferable to ultra-optimized but undebuggable code. Here it is simply a triangular walk over a square array so I wouldn't expect too much discrepancy.

**Comment by epsilonmach** [comment]
@ilayn I cannot seem to access the paper you have referenced (DOI:10.1007/BF01933218), even through a university-based journal access portal. Would you happen to know of any non-paywall URL for the paper? Alternatively, do you know if the proposed algorithm is the same as in the Alexander et al paper referenced above?

**Comment by ilayn** [comment]
@epsilonmach Here is the excerpt for the algo I've coded above by removing the internal loop by NumPy ufuncs. 

![image](https://user-images.githubusercontent.com/1303842/34323124-606b3cb8-e83b-11e7-806d-646c9dcf21f6.png)

As I mentioned above this is also written by Pan who is also the 2nd author of the Alexander et al paper. The author claims this simplifies the hyperbolic rotation method (aka LINPACK code). 

Even proposes another algo C but that didn't seem to make too much of a difference though being more complicated. This one however is, obviously, pretty easy to code.

**Comment by epsilonmach** [comment]
@ilayn That was very helpful -- thank you!

**Comment by ilayn** [comment]
@ashimb9 Are you working on this currently? I'm asking just just to understand for myself and for other potential contributors. Definitely not to put pressure, (ok well maybe a tiny bit) 😃

**Comment by ashimb9** [comment]
@ilayn Hi, yes. Sorry I was not able to work on it much due to the holidays but I will be able to pick it up again by the end of this week. I hope that is okay?

**Comment by ilayn** [comment]
Sure no problem.

**Comment by mdhaber** [comment]
@ilayn gh-16499 says that it follows [Pan 1990](https://link.springer.com/article/10.1007/BF01933218) for rank-1 updates and downdates and [Van De Geijn 2011](https://dl.acm.org/doi/abs/10.1145/2049662.2049666) for rank-p updates and downdates.

However, in gh-8286, you stated that 
>  downdate and update uses different strategies namely, update is Krause Igel 2005, and downdate is Pan's algorithm that was discussed in [gh-8188]

This suggested that there is some way that these things are always done / must be done. When asked why, you pointed to [Julia's code](https://github.com/JuliaLang/julia/blob/23132607d0b87a2cdd59245dc70eb7aa905df26b/stdlib/LinearAlgebra/src/cholesky.jl#L888-L1030) as a reference. You also seemed to indicate that [the code you wrote above](https://github.com/scipy/scipy/issues/8188#issuecomment-351902773) should _not_ be used.

Do I understand that a requirement for addressing gh-8188 is that we reproduce Julia's strategy for updates/downdates in SciPy, or are other algorithms allowable provided that they meet certain goals? If so, what are those goals that were not met by gh-8286? 

Please note that gh-8286 already satisfied these three criteria:

- Produce a valid updated Cholesky factor when it exists.
- Detect when the updated matrix would not be positive definite.
- Exceed the performance of factorizing the updated matrix from scratch.
